elements/lisk-p2p/src/peer_directory/peer_list.ts

```diff
@@ -190,23 +185,23 @@ export class PeerList {
 		return result;
 	}
 
-	public evictPeer(bucketId: number): CustomPeerInfo {
+	public evictPeer(bucketId: number): CustomPeerInfo | undefined {
 		return this.evictRandomlyFromBucket(bucketId);
 	}
 	// If there are no peers which are old enough to be evicted based on number of days then pick a peer randomly and evict.
-	protected evictRandomlyFromBucket(bucketId: number): CustomPeerInfo {
-		const peerList = this.peerMap.get(bucketId);
-		if (!peerList) {
-			throw new Error(`No Peers exist for bucket Id: ${bucketId}`);
+	protected evictRandomlyFromBucket(
+		bucketId: number,
+	): CustomPeerInfo | undefined {
+		const bucket = this.peerMap.get(bucketId);
+		if (!bucket) {
+			return undefined;
```

Comment:
I think we should throw an error in this case because it is not a possible scenario and make this function always return an evictedPeer. By code, a bucket must exist only then the eviction can be triggered but TS always think `Map.get` can return `undefined` which is not true in this case

Reply:
It would be a problem if evict was called on an empty bucket.

Before, if that happened, it would throw an error https://github.com/LiskHQ/lisk-sdk/pull/4292/files#diff-29e3f653688df876f4d60b83818a6cd8L200 - But that error was not handled anywhere so it adds risk.

If the function tried to evict a specific peer based on a peer id, then it would make sense to throw an error but in this case we evict based on the bucket ID so it's OK to make that operation fail-safe if the bucket is empty.

Reply:
I agree that we need to then handle the error everywhere. I have one more solution to make this function always return a peer.
As we already check if the bucket exists or not in `addPeer` function, We can pass `bucket` instead of `bucketId` from `addPeer` function to the `evictPeer` function. I think we can reduce some computation as well to compute the bucketId again and again